docs/site/Life-cycle.md

```diff
@@ -0,0 +1,217 @@
+}
+```
+
+## Discover life cycle observers
+
+The `Application` finds all bindings tagged with `CoreTags.LIFE_CYCLE_OBSERVER`
+within the context chain and resolve them as observers to be notified.
+
+## Notify life cycle observers of start/stop related events by order
+
+There may be dependencies between life cycle observers and their order of
+processing for `start` and `stop` need to be coordinated. For example, we
+usually start a server to listen on incoming requests only after other parts of
+the application are ready to handle requests. The stop sequence is typically
+processed in the reverse order. To support such cases, we introduce
+two-dimension steps to control the order of life cycle actions.
+
+### Observer groups
+
+First of all, we allow each of the life cycle observers to be tagged with a
+group. For example:
+
+- datasource (connect/disconnect)
+
+  - mongodb
+  - mysql
+
+- server
+  - rest
+  - gRPC
+
+We can then configure the application to trigger observers group by group as
+configured by an array of groups in order such as `['datasource', 'server']`.
+Observers within the same group can be notified in parallel.
+
+For example,
+
+```ts
+app
+  .bind('observers.MyObserver')
+  .toClass(MyObserver)
+  .tag({
+    [CoreTags.LIFE_CYCLE_OBSERVER_GROUP]: 'g1',
+  })
+  .apply(asLifeCycleObserver);
+```
+
+The observer class can also be decorated with `@bind` to provide binding
+metadata.
+
+```ts
+import {bind, createBindingFromClass} from '@loopback/context';
+import {CoreTags, asLifeCycleObserver} from '@loopback/core';
+
+@bind(
+  {
+    tags: {
+      [CoreTags.LIFE_CYCLE_OBSERVER_GROUP]: 'g1',
+    },
+  },
+  asLifeCycleObserver,
+)
+export class MyObserver {
+  // ...
+}
+
+app.add(createBindingFromClass(MyObserver));
+```
+
+The order of observers are controlled by a `groupsByOrder` property of
+`LifeCycleObserverRegistry`, which receives its options including the
+`groupsByOrder` from `CoreBindings.LIFE_CYCLE_OBSERVER_OPTIONS`. Thus the
+initial `groupsByOrder` can be set as follows:
+
+```ts
+app
+  .bind(CoreBindings.LIFE_CYCLE_OBSERVER_OPTIONS)
+  .to({groupsByOrder: ['g1', 'g2', 'server']});
+```
+
+Or:
+
+```ts
+const registry = await app.get(CoreBindings.LIFE_CYCLE_OBSERVER_REGISTRY);
+registry.setGroupsByOrder(['g1', 'g2', 'server']);
+```
+
+Observers are sorted using `groupsByOrder` as the relative order. If an observer
+is tagged with a group that are not in `groupsByOrder`, it will come before any
+groups within `groupsByOrder`. Such custom groups are also sorted by their names
+alphabetically.
+
+In the example below, `groupsByOrder` is set to `['g1', 'g2']`. Given the
+following observers:
+
+- 'my-observer-1' ('g1')
+- 'my-observer-2' ('g2')
+- 'my-observer-4' ('2-custom-group')
+- 'my-observer-3' ('1-custom-group')
+
+The sorted observer groups will be:
+
+```ts
+{
+  '1-custom-group': ['my-observer-3'],
+  '2-custom-group': ['my-observer-4'],
+  'g1': ['my-observer-1'],
+  'g2': ['my-observer-2'],
```

Comment:
This example would work for multiple different sorting rules. For example, `1-custom-group` is alphabetically preceding `g1`, also `g1` is before `g2`. Your entire list is sorted in the alphabetical order, it's not clear which part is controller by specific rules for unknown groups, which part is controlled by `groupsByOrder` and which is controlled by alphabetical sorting.


I am proposing to use the following example instead:

- `groupsByOrder `is set to `['setup-servers', 'announce-services']`
- 'my-observer-1' in group 'setup-servers'
- 'my-observer-2' in group 'announce-services'
- 'my-observer-4' in 'my-custom-group'
- 'my-observer-3' in 'another-custom-group'

Also please check how is this section rendered from Markdown to HTML, I think it may need a bit of polishment.